urllib3/util/retry.py

```diff
@@ -290,5 +290,31 @@ def __repr__(self):
                     cls=type(self), self=self)
 
 
+def handle_redirect(response, requester, retries, method, url, *args, **kwargs):
+    """
+    Abstraction of the redirect process so that we can access it from
+    multidomain levels in addition to from the :class:`.ConnectionPool`.
+    """
+    # urljoin will give us either a full URL or just a path, based on what
+    # the original URL is, making it appropriate for the ConnectionPool
+    # level as well as the PoolManager or SessionManager
+    redirect_location = urljoin(url, response.get_redirect_location())
```

Comment:
At a minimum, it handles all the cases that are currently tested for on `ConnectionPool` and `PoolManager`.

I'm not clear why it wasn't used across the board in the past; it was present in `PoolManager`'s block of redirect logic, but not in `ConnectionPool`'s. I suspect that it may have been discovered when setting up the cross-domain redirect logic for `PoolManager`, but wasn't applied to `ConnectionPool` because at the time they were independent pieces of logic.

In any manual tests I do, it seems to behave as I'd hope, including with locations not containing a scheme. Essentially:

    >>> urljoin('http://host/path/resource', 'http://other_host/path/resource')
    'http://otherhost/path/resource'
    >>> urljoin('http://host/path/resource', '/otherpath/otherresource')
    'http://host/otherpath/otherresource'
    >>> urljoin('http://host/path/resource', 'otherresource')
    'http://host/path/otherresource'
    >>> urljoin('/path/resource', '/otherresource')
    '/otherresource'
    >>> urljoin('/path/resource', 'nested/otherresource')
    '/path/nested/otherresource'